processor/src/advice/mod.rs

```diff
@@ -19,38 +19,29 @@ pub use source::AdviceSource;
 // ADVICE PROVIDER
 // ================================================================================================
 
-// TODO elaborate on why this is non-deterministic (seems it is a simple state machine).
-
-// TODO the `advance_clock` seems to be an internal of the VM and shouldn't necessarily be here.
-
-// TODO this will likely suffer a breaking change as we introduce a generic storage for big Merkle
-// sets.
-// The purpose of this clock is to keep the feedback of the provided in sync with the execution
-// trace of the VM, but on the other hand the VM can control the calls/traces itself, without the
-// need for the advice provider (or any other external component) to keep track of it. If we
-// delegate this control to the advice provider - especially if it is a trait, the user might
-// implement it incorrectly, creating undefined behavior on the VM side (that expects the counter
-// to incrementally increase).
-
-/// Common behavior of advice providers for program execution.
+/// Defines behavior of an advice provider.
 ///
-/// An advice provider supplies non-deterministic inputs to the processor.
+/// An advice provider is a component through which the VM processor can interact with the host
+/// environment. The processor can request nondeterministic inputs from the advice provider (i.e.,
+/// result of a computation performed outside of the VM), as well as insert new data into the
+/// advice provider.
 ///
-/// 1. Provide a stack functionality that yields elements as a stack (last in, first out). These can
-///    be yielded as elements, words or double words.
-/// 2. Provide a map functionality that will store temporary stacks that can be appended to the main
-///    stack. This operation should not allow key overwrite; that is: if a given key exists, the
-///    implementation should error if the user attempts to insert this key again, instead of the
-///    common behavior of the maps to simply override the previous contents. This is a design
-///    decision to increase the runtime robustness of the execution.
-/// 3. Provide merkle tree interfaces, backed by a [MerkleStore].
+/// An advice provider consists of the following components:
+/// 1. Advice stack, which is a LIFO data structure. The processor can move the elements from the
```

Comment:
nit: we can mention it's a LIFO, but this is implicit as it will be always the case, unless otherwise specified.